src/api/v1/db/oracledb/pizzas-dao.js

```diff
@@ -217,4 +248,104 @@ const getPizzaById = async (pizzaId, query) => {
+ * `body.data.relationships.ingredients`
+ * @returns {object[]} an object with `queries` and `bindParams` members
+ */
+const getIngredientQueriesAndBindParams = async (ingredientRelationshipData) => {
+  const bindParams = {};
+  const ingredientIds = ingredientRelationshipData
+    .map((ingredient) => ingredient.id);
+  if (!await checkIngredientsExist(ingredientIds)) {
+    throw new Error('ingredient IDs are invalid');
+  }
+  const queries = ingredientIds.map((id) => {
+    bindParams[`ingredientId${id}`] = parseInt(id, 10);
+    return dedent`
+    INTO PIZZA_INGREDIENTS (INGREDIENT_ID, PIZZA_ID)
+      VALUES (:ingredientId${id}, :pizzaId)
+    `;
+  });
+  return { queries, bindParams };
+};
+
+/**
+ * insert ingredients relationships resources into the
+ * PIZZA_INGREDIENTS table
+ *
+ * @param {string[]} ingredientQueries
+ * @param {object} ingredientBindParams
+ * @param {OracleDB.Connection} connection
+ */
+const insertIngredients = async (ingredientQueries, ingredientBindParams, connection) => {
+  const insertIngredientsQuery = dedent`
+  INSERT ALL
+    ${ingredientQueries.join('\n  ')}
+  SELECT * FROM dual
+  `;
+  await connection.execute(
+    insertIngredientsQuery,
+    ingredientBindParams,
+    { autoCommit: true },
+  );
+};
+/**
+ * Create a pizza record in the database
+ *
+ * @param {object} body
+ * @returns {Promise<object>} the serialized, newly created pizza
+ */
+const postPizza = async (body) => {
+  const pizzaBindParams = getBindParams(body, pizzaProperties, pizzaOutBindParams);
+  const connection = await getConnection();
+  let ingredientQueries = [];
+  let ingredientBindParams = {};
+
+  const hasIngredients = 'relationships' in body.data && 'ingredients' in body.data.relationships;
+
+  if (hasIngredients) {
+    try {
+      const { queries, bindParams } = await getIngredientQueriesAndBindParams(
+        body.data.relationships.ingredients.data,
+      );
+      ingredientQueries = queries;
+      ingredientBindParams = bindParams;
+    } catch (e) {
+      if (e.message === 'ingredient IDs are invalid') return null;
+      throw e;
+    }
+  }
+
+  pizzaBindParams.doughId = _.get(body.data, 'relationships.dough.data.id', null);
+
+  try {
+    const result = await connection.execute(
+      pizzaPostQuery,
+      pizzaBindParams,
+      { autoCommit: true },
+    );
+
+    const rawPizza = convertOutBindsToRawResource(result.outBinds);
+
+    if (hasIngredients) {
+      ingredientBindParams.pizzaId = rawPizza.id;
+      await insertIngredients(ingredientQueries, ingredientBindParams, connection);
+    }
+
+    return serializePizza(rawPizza, 'pizzas');
+  } catch (e) {
+    if ('errorNum' in e && e.errorNum === 2291) {
+      // oracleDB integrity constraint -- tried to insert invalid primary key for dough
+      return null;
+    }
+
+    throw new Error(`unhandled exception: ${e.message}`);
+  } finally {
+    connection.close();
+  }
```

Comment:
steps here are basically:
1. check if ingredient IDs are valid
2. insert pizza resource
3. insert ingredient resources

We could merge these into one step if it was possible to write the query in such a way that we could get the ID of the pizza as we were creating it (I don't think this is possible) AND guarantee that all of the insertions would fail if the ingredient IDs used were invalid. `INSERT ALL` solves the second problem, but not the first, and using `connection.executeMany` could potentially solve the first problem but doesn't solve the second one.

I'm very interested to hear if there's a better way to solve this, but this is the best I could come up with.

Reply:
I can't think of a better way to handle this. It should be fine to leave it how it is for now.